cci_cameroon/analysis/predictive_model/model_run_report.py

```diff
@@ -0,0 +1,186 @@
+
+# %%
+y_train = np.concatenate((y_train, y_nr_train))
+y_test = np.concatenate((y_test, y_nr_test))
+
+# %%
+X_train = X_train.append(no_response_train, ignore_index=False)
+X_test = X_test.append(no_response_test, ignore_index=False)
+
+# %%
+X_train, y_train = shuffle(X_train, y_train, random_state=1)
+X_test, y_test = shuffle(X_test, y_test, random_state=1)
+
+# %%
+# Get language model
+model_fr = SentenceTransformer("Sahajtomar/french_semantic")  # French language model
+# Encode train and test transform into word embeddings
+X_train_embeddings_fr = model_fr.encode(list(X_train))
+X_test_embeddings_fr = model_fr.encode(list(X_test))
+
+# %% [markdown]
+# ### Fit and predict with best performing models
+
+# %%
+# Best performing models
+svm = MultiOutputClassifier(SVC(C=3, gamma="scale"), n_jobs=-1)
+knn = KNeighborsClassifier(n_neighbors=5, p=1, weights="distance")
+rf = RandomForestClassifier(n_estimators=200, random_state=1)
+dt = DecisionTreeClassifier(criterion="entropy", random_state=1)
+nb = MultiOutputClassifier(GaussianNB(), n_jobs=-1)
+
+# %%
+# Fit and predict on test set
+svm.fit(X_train_embeddings_fr, y_train)
+y_pred_svm = svm.predict(X_test_embeddings_fr)
+knn.fit(X_train_embeddings_fr, y_train)
+y_pred_knn = knn.predict(X_test_embeddings_fr)
+rf.fit(X_train_embeddings_fr, y_train)
+y_pred_rf = rf.predict(X_test_embeddings_fr)
+dt.fit(X_train_embeddings_fr, y_train)
+y_pred_dt = dt.predict(X_test_embeddings_fr)
+nb.fit(X_train_embeddings_fr, y_train)
+y_pred_nb = nb.predict(X_test_embeddings_fr)
+
+# %%
+# Update test set and predictions to include class for 'no class'
+y_test = mtr.add_y_class(y_test)
+y_pred_svm = mtr.add_y_class(y_pred_svm)
+y_pred_knn = mtr.add_y_class(y_pred_knn)
+y_pred_rf = mtr.add_y_class(y_pred_rf)
+y_pred_dt = mtr.add_y_class(y_pred_dt)
+y_pred_nb = mtr.add_y_class(y_pred_nb)
+
+# %%
+codes.append("Not classified as any of the eight codes")
+
+# %%
+# Create confusion matrix from the best performing models
+cm_svm = multilabel_confusion_matrix(y_test, y_pred_svm)
+cm_knn = multilabel_confusion_matrix(y_test, y_pred_knn)
+cm_rf = multilabel_confusion_matrix(y_test, y_pred_rf)
+cm_dt = multilabel_confusion_matrix(y_test, y_pred_dt)
+cm_nb = multilabel_confusion_matrix(y_test, y_pred_nb)
+
+# %%
+# %%capture
+mtr.save_cm_plots(cm_svm, "svm", codes)
+mtr.save_cm_plots(cm_knn, "knn", codes)
+mtr.save_cm_plots(cm_rf, "random_forest", codes)
+mtr.save_cm_plots(cm_dt, "decision_tree", codes)
+mtr.save_cm_plots(cm_nb, "naive_bayes", codes)
+
+# %% [markdown]
+# ### Common words in each predicted class
+# Using the predictions from the KNN model looking at the most common words in each predicted class and saving the results as bar charts.
+
+# %%
+# Remove last class 'no code'
+codes = codes[:-1]
+
+# %%
+pred_proba = knn.predict_proba(X_train_embeddings_fr)
+
+# %%
+counts = mtr.word_counts_class(codes, pred_proba, X_train, stop)
+
+# %%
+mtr.common_words_plots(codes, counts)
+
```

Comment:
Can see figures being saved but not the unclassified data which needs to be fed into the clustering algorithm. Having the data stored in a dataframe with a similar structure as the data in model_data variable would make this compatible with the format used in the clustering exercise. So functions written would be able to access a column  from read data and process it accordingly.